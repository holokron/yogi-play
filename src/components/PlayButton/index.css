.btn-rounded {
    border-radius: 30px !important;
}

.play-btn {
    font-size: 13px !important;    
}