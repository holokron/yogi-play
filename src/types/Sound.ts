export default interface Sound {
  id: string;
  name: string;
  path: string;
  tags?: object;
  order?: number;
  isNew?: boolean;
}
