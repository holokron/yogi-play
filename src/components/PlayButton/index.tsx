import * as React from 'react'
import Button from 'reactstrap/lib/Button'
import ButtonGroup from 'reactstrap/lib/ButtonGroup'
import './index.css'
import Sound from '../../types/Sound'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import Col from 'reactstrap/lib/Col';

interface ButtonWrapperProps {
    children: React.ReactNode
}

const ButtonWrapper: React.FunctionComponent<ButtonWrapperProps> = React.memo(({ children }: ButtonWrapperProps) => (
    <Col
        xl="2"
        lg="2"
        md="3"
        sm="4"
        xs="6"
        className="mb-2 pl-1 pr-1"
    >
        {children}
    </Col>
))

export interface Props {
    sound: Sound
    onClick?: () => void
    onFavClick?: () => void
    isPlaying?: boolean
    isLoading?: boolean
    isFavourite?: boolean
}

export default class PlayButton extends React.PureComponent<Props, {}> {
    public static defaultProps: Partial<Props> = {
        isPlaying: false,
        isLoading: false,
        isFavourite: false,
    }

    public renderPlayButton() {
        const {
            sound,
        } = this.props

        return (
            <Button {...this.getPlayButtonProps()}>
                {sound.name}
            </Button>
        )
    }

    public renderFavouriteButton() {
        return (
            <Button {...this.getFavButtonProps()}>
                <FontAwesomeIcon icon="star" />
            </Button>
        )
    }

    public render() {
        return (
            <ButtonWrapper>
                <ButtonGroup {...this.getButtonGroupProps()}>
                    {this.renderPlayButton()}
                    {this.renderFavouriteButton()}
                </ButtonGroup>
            </ButtonWrapper>
        )
    }

    private getButtonGroupProps(): object {
        const {
            isLoading,
            isPlaying,
        } = this.props

        let animationClass: string = ''

        if (isLoading) {
            animationClass += 'animated infinite flash'
        }

        if (isPlaying) {
            animationClass += 'animated infinite pulse active'
        }

        return {
            className: `play-btn d-flex ${animationClass}`,
            size: 'sm',
        }
    }

    private getPlayButtonProps(): object {
        const {
            onClick,
        } = this.props

        return {
            className: 'play-btn__play text-truncate text-uppercase font-weight-bold',
            color: 'primary',
            outline: true,
            onClick,
        }
    }

    private getFavButtonProps(): object {
        const {
            isFavourite,
            onFavClick,
        } = this.props

        return {
            className: 'play-btn__fav',
            color: 'primary',
            outline: !isFavourite,
            onClick: onFavClick,
        }
    }
}