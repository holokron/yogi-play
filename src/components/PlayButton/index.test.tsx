import * as React from "react";
import * as renderer from "react-test-renderer";
import { mount } from "enzyme";
import PlayButton from ".";
import Sound from "../../types/Sound";

describe("@components/PlayButton", () => {
  it("should match snapshot", () => {
    const sound: Sound = {
      id: "test",
      name: "Test",
      path: "/foo/bar.mp3"
    };

    const tree = renderer.create(<PlayButton sound={sound} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should match snpashot when audio is playing", () => {
    const sound: Sound = {
      id: "test",
      name: "Test",
      path: "/foo/bar.mp3"
    };

    const tree = renderer
      .create(<PlayButton sound={sound} isPlaying />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should have correct props", () => {
    const sound: Sound = {
      id: "test",
      name: "Test",
      path: "/foo/bar.mp3"
    };

    const isPlaying = false;

    const result = mount(<PlayButton sound={sound} isPlaying={isPlaying} />);

    expect(result.props().sound).toEqual(sound);
    expect(result.props().isPlaying).toEqual(isPlaying);
    expect(result.props().onClick).toBeUndefined();
  });

  it("should execute onClick", () => {
    const sound: Sound = {
      id: "test",
      name: "Test",
      path: "/foo/bar.mp3"
    };

    let changeAfterClick = false;

    const onClick = (): void => {
      changeAfterClick = true;
    };

    const result = mount(<PlayButton sound={sound} onClick={onClick} />);

    result.find("button.play-btn__play").simulate("click");

    expect(changeAfterClick).toEqual(true);
  });
});
