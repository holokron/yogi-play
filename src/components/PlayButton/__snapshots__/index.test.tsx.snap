// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@components/PlayButton should match snapshot 1`] = `
<div
  className="mb-2 pl-1 pr-1 col-6 col-sm-4 col-md-3 col-lg-2 col-xl-2"
>
  <div
    className="play-btn d-flex  btn-group-sm btn-group"
    role="group"
  >
    <button
      aria-label={null}
      className="play-btn__play text-truncate text-uppercase font-weight-bold btn btn-outline-primary"
      onClick={[Function]}
    >
      Test
    </button>
    <button
      aria-label={null}
      className="play-btn__fav btn btn-outline-primary"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-star fa-w-18 "
        data-icon="star"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`@components/PlayButton should match snpashot when audio is playing 1`] = `
<div
  className="mb-2 pl-1 pr-1 col-6 col-sm-4 col-md-3 col-lg-2 col-xl-2"
>
  <div
    className="play-btn d-flex animated infinite pulse active btn-group-sm btn-group"
    role="group"
  >
    <button
      aria-label={null}
      className="play-btn__play text-truncate text-uppercase font-weight-bold btn btn-outline-primary"
      onClick={[Function]}
    >
      Test
    </button>
    <button
      aria-label={null}
      className="play-btn__fav btn btn-outline-primary"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-star fa-w-18 "
        data-icon="star"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
  </div>
</div>
`;
